docs/references/ingress-messages.mdx

```diff
@@ -0,0 +1,84 @@
+
+### Recommendations and best practices for secure update call handling
+
+The best way to handle the aforementioned errors is to implement safe retries using idempotent calls. Idempotency and several patterns to achieve it are discussed in the [best practices on safe retries and idempotency](/docs/current/developer-docs/smart-contracts/best-practices/idempotency).
+
+It may not always be possible to make call endpoints idempotent. For example, an endpoint that is not under the developer’s control may simply not have this property. In this case, there is still a secure way to retry the calls by relying on built-in deduplication based on request ids, as we describe below. However, the guarantees in this case are weaker compared to using idempotent calls.
+
+ICP provides a built-in deduplication feature. Namely, if two calls with identical [request id](/docs/current/references/ic-interface-spec#request-ids) are made, the message is processed at most once. For two calls to have the same `request_id`, they must have the same values for all fields `request_type`, `sender`, `nonce`, `canister_id`, `method_name`, `arg` and **ingress_expiry**. This gives idempotency, but only until the [`ingress_expiry`](/docs/current/references/ic-interface-spec#authentication) is reached, at which point the message expires and ICP [may remove the `request_id` from the system state tree](/docs/current/references/ic-interface-spec#state-tree-request-status). This isn't an issue since the same request will not be accepted later, as it expired. Thus, such requests can be safely retried analogously to how it is done for [idempotent calls](/docs/current/developer-docs/smart-contracts/best-practices/idempotency#idempotent-canister-apis), see the illustration on that page.
+
+As ICP rejects `ingress_expiry` that is too far in the future, the time window for retrying is limited. To maximize the window, one can maximize `ingress_expiry` in this case and time out the retries after `ingress_expiry` as well.
+
+There are two options for retries:
+* Only poll `read_state` until `ingress_expiry` is reached. Note that this may be hard if the IC time and local time differ. This is currently the behaviour of agent-js and agent-rs.
+* Poll `read_state` for e.g. `30s` and then retry the `call`, loop until `ingress_expiry` (e.g. `5min`). This can be beneficial because a call that didn’t make it would be retried.
+
+If the expiry is reached, there is no way for the client to determine if the call was successful unless the canister’s API provides this information. In this case, after the message has expired, the user must be instructed to manually check the status of the call using application-specific APIs. There can be (hopefully uncommon) cases where the APIs don't allow checking this, in which case the user simply can’t tell if the message has been executed. Ideally, applications set expiry values such that under typical load timeouts are rare.
```

Comment:
This is not entirely correct, expiry only determines when the canister execution can start. If processing takes 10min, then the response will be available after 10min for another 5min (assuming there is space in the ingress history, if not then it does not hold even before expiry is reached, iirc).

Maybe rephrase as follows:

```suggestion
Some time after expiry is reached, there is no way for the client to determine if the call was successful unless the canister’s API provides this information. In this case, the client can check the status of the call using application-specific APIs. There can be (hopefully uncommon) cases where the APIs don't allow checking this, in which case the user simply can’t tell if the message has been executed. Ideally, applications set expiry values such that under typical load timeouts are rare.
```

When do you use "user" and when "client"? Imho, checking the status does not need to be manual. If it's crucial for the app behavior, then the client/user-side code should do this automatically

Reply:
Thanks, applied the suggestion!

> When do you use "user" and when "client"? Imho, checking the status does not need to be manual. If it's crucial for the app behavior, then the client/user-side code should do this automatically

"user" indicates the end user, and "client" can also be code such as an agent. I think here it's better to talk about the client. Changed "user" to "client" in the sentence above.